### PR TITLE
[SPARK-14224] [SPARK-14223] [SPARK-14310] [SQL] fix RowEncoder and parquet reader for wide table

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

1) fix the RowEncoder for wide table (many columns) by splitting the generate code into multiple functions.
2) Separate DataSourceScan as RowDataSourceScan and BatchedDataSourceScan
3) Disable the returning columnar batch in parquet reader if there are many columns.
4) Added a internal config for maximum number of fields (nested) columns supported by whole stage codegen.

Closes #12098 
## How was this patch tested?

Add a tests for table with 1000 columns.

### CODE REVIEW

**Comment by davies** [comment]
cc @nongli @rxin @marmbrus

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 1]
```diff
@@ -224,9 +224,13 @@ public ColumnarBatch resultBatch() {
 
   /**
    * Can be called before any rows are returned to enable returning columnar batches directly.
+   *
+   * @param maxColumns The maximum number of fields supported by whole stage codegen.
    */
-  public void enableReturningBatches() {
-    returnColumnarBatch = true;
+  public boolean tryEnableReturningBatches(int maxColumns) {
```
Comment (marmbrus) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Why are we pushing a config value into the reader so that it can switch?  Can't we just put this logic in the query planner where the decision to call `enableReturningBatches` is.  I think its clearer if there is a single place where we decide to use column batches or not.

Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Yes, we should do this in planner, by checking the schema, which requires some cleanup in parquet reader, currently which still use runtime exception to do the switch. cc @nongli @sameeragarwal 

Should we wait for that?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 2]
```diff
@@ -121,6 +109,8 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
       throws IOException, InterruptedException, UnsupportedOperationException {
     super.initialize(inputSplit, taskAttemptContext);
     initializeInternal();
+    Configuration conf = ContextUtil.getConfiguration(taskAttemptContext);
+    returnColumnarBatch = conf.getBoolean("returning.batch", false);
```
Comment (rxin) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: where is this returning.batch set?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala [hunk 3]
```diff
@@ -175,10 +175,11 @@ private[sql] case class DataSourceScan(
   }
 
   private def canProcessBatches(): Boolean = {
+    lazy val conf = SQLContext.getActive().get.conf
     relation match {
       case r: HadoopFsRelation if r.fileFormat.isInstanceOf[ParquetSource] &&
-        SQLContext.getActive().get.conf.getConf(SQLConf.PARQUET_VECTORIZED_READER_ENABLED) &&
-        SQLContext.getActive().get.conf.getConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED) =>
+        conf.parquetVectorizedReaderEnabled &&
+        conf.wholeStageEnabled && schema.length <= conf.wholeStageMaxNumFields =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: why do we need to disable batching if num fields is large?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 4]
```diff
@@ -434,12 +435,20 @@ case class CollapseCodegenStages(conf: SQLConf) extends Rule[SparkPlan] {
...
+    case _ => 1
+  }
+
   private def supportCodegen(plan: SparkPlan): Boolean = plan match {
     case plan: CodegenSupport if plan.supportCodegen =>
       val willFallback = plan.expressions.exists(_.find(e => !supportCodegen(e)).isDefined)
       // the generated code will be huge if there are too many columns
-      val haveManyColumns = plan.output.length > 200
-      !willFallback && !haveManyColumns
+      val haveManyFields = numOfNestedFields(plan.schema) > conf.wholeStageMaxNumFields
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: haveTooManyFields

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala (hunk 3 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: is it because we have to synchronize this with whole stage codegen because we don't have great planner support for this?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: We really need to move all of this logic into the planner.  I don't think it would be that hard after the refactoring we did.

I think main complication is that the parquet reader only supports a subset of possible schemata.  So you have to branch based both on the format but also arbitrary functions of the schema.

**Comment by davies** [comment]
@marmbrus Did another round of refactor, does this match the things in your mind?

**Comment by marmbrus** [comment]
Yeah, this looks much cleaner.

**Comment by yhuai** [comment]
@marmbrus Is this PR good to go?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 5]
```diff
@@ -481,6 +497,17 @@ private[sql] class DefaultSource
   }
 }
 
+/**
+ * The ParquetInputFormat that create VectorizedParquetRecordReader.
+ */
+final class VectorizedParquetInputFormat extends ParquetInputFormat[InternalRow] {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: @liancheng @cloud-fan lets make sure we delete this when you remove `buildInternalScan`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 6]
```diff
@@ -408,6 +411,10 @@ private[sql] class DefaultSource
     val parquetFilterPushDown = sqlContext.conf.parquetFilterPushDown
     val assumeBinaryIsString = sqlContext.conf.isParquetBinaryAsString
     val assumeInt96IsTimestamp = sqlContext.conf.isParquetINT96AsTimestamp
+    // TODO: this could be wrong iff:
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: What's up with this TODO?  Does this matter if we get rid of buildInternalScan?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 7]
```diff
@@ -306,6 +315,10 @@ private[sql] class DefaultSource
       SQLConf.PARQUET_INT96_AS_TIMESTAMP.key,
       sqlContext.conf.getConf(SQLConf.PARQUET_INT96_AS_TIMESTAMP))
 
+    // Whole stage codegen (PhysicalRDD) is able to deal with batches directly
+    parquetConf.setBoolean(ParquetRelation.RETURNING_BATCH,
+      supportBatch(sqlContext, StructType(partitionSchema.fields ++ dataSchema.fields)))
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: In general, I'd prefer if all new code passes options explicitly to the closure instead of through the conf.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 8]
```diff
@@ -469,6 +469,13 @@ trait FileFormat {
       options: Map[String, String]): RDD[InternalRow]
 
   /**
+   * Returns whether this format support returning columnar batch or not.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Maybe a TODO, that we should just have different traits for the different formats.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala [hunk 9]
```diff
@@ -279,7 +279,8 @@ class FileSourceStrategySuite extends QueryTest with SharedSQLContext with Predi
   /** Plans the query and calls the provided validation function with the planned partitioning. */
   def checkScan(df: DataFrame)(func: Seq[FilePartition] => Unit): Unit = {
     val fileScan = df.queryExecution.executedPlan.collect {
-      case DataSourceScan(_, scan: FileScanRDD, _, _) => scan
+      case scan: DataSourceScan if scan.rdd.isInstanceOf[FileScanRDD] =>
+        scan.rdd.asInstanceOf[FileScanRDD]
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala]: Why this change?  Pattern matching seems cleaner to me.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 10]
```diff
@@ -396,6 +396,12 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val WHOLESTAGE_MAX_NUM_FIELDS = SQLConfigBuilder("spark.sql.codegen.maxFields")
+    .internal()
+    .doc("The maxinum number of fields (nested) could be supported by whole stage codegen")
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Nit: The maximum number of fields (including nested fields) that will be supported before deactivating whole-stage codegen.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 11]
```diff
@@ -347,32 +358,24 @@ private[sql] class DefaultSource
...
 
-      val parquetReader = try {
-        if (!enableVectorizedParquetReader) sys.error("Vectorized reader turned off.")
+      val parquetReader = if (enableVectorizedParquetReader) {
         val vectorizedReader = new VectorizedParquetRecordReader()
         vectorizedReader.initialize(split, hadoopAttemptContext)
         logDebug(s"Appending $partitionSchema ${file.partitionValues}")
         vectorizedReader.initBatch(partitionSchema, file.partitionValues)
-        // Whole stage codegen (PhysicalRDD) is able to deal with batches directly
         // TODO: fix column appending
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: I think this TODO is stale.  We are appending the columns on the line above.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 12]
```diff
@@ -347,32 +358,24 @@ private[sql] class DefaultSource
...
-            case Some(filter) =>
-              new ParquetRecordReader[InternalRow](
-                new CatalystReadSupport,
-                FilterCompat.get(filter, null))
-            case _ =>
-              new ParquetRecordReader[InternalRow](new CatalystReadSupport)
-          }
-          reader.initialize(split, hadoopAttemptContext)
-          reader
+      } else {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: I think we lost the log messages that tell us that we are falling back on a different parquet reader.

I love that we are getting rid of the try/catch!

**Comment by marmbrus** [comment]
This is a huge improvement.   A few minor comments, otherwise LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala (hunk 9 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala]: DataSourceScan is a trait right now

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala (hunk 6 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Once we get rid of buildInternalScan(), all of these will be deleted.

**Comment by davies** [comment]
@marmbrus Since we will remove buildInternalScan soon, removed the returning batch support for that case to simplify this patch.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 13]
```diff
@@ -121,6 +109,7 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
       throws IOException, InterruptedException, UnsupportedOperationException {
     super.initialize(inputSplit, taskAttemptContext);
     initializeInternal();
+    Configuration conf = ContextUtil.getConfiguration(taskAttemptContext);
```
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: what does this do?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 14]
```diff
@@ -434,12 +435,20 @@ case class CollapseCodegenStages(conf: SQLConf) extends Rule[SparkPlan] {
     case _ => true
   }
 
+  private def numOfNestedFields(dataType: DataType): Int = dataType match {
+    case dt: StructType => dt.fields.map(f => numOfNestedFields(f.dataType)).sum
+    case m: MapType => numOfNestedFields(m.keyType) + numOfNestedFields(m.valueType)
+    case a: ArrayType => numOfNestedFields(a.elementType) + 1
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: why + 1

**Comment by nongli** [comment]
LGTM

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java (hunk 13 above)
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Not needed, will remove

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala (hunk 14 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: Considering generate a UnsafeRow, ArrayType itself will need some code to assemble the elements, co counting it as as `1`.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: we should rename this function then. it is really not about the number of fields? It's some estimation of local variables in codegen.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: Removed this `+1`, make it consistent with StructType and MapType.

**Comment by davies** [comment]
Since the last commit is tiny, merging this into master to unblock others.
